lib/api.rb

```diff
@@ -243,12 +262,129 @@ def validation_error(msg)
+	  response['access-control-allow-origin'] = '*'
+    d = device_from_bearer
+    if !d
+      return validation_error("invalid bearer")
+    end
+	d.user.private_key = params[:encryptedPrivateKey]
+	d.user.public_key = params[:publicKey]
+
+    {
+	  "Id" => d.user_uuid,
+	  "Name" => d.user.name,
+	  "Email" => d.user.email,
+	  "EmailVerified" => d.user.email_verified,
+	  "Premium" => d.user.premium,
+	  "MasterPasswordHint" => d.user.password_hint,
+	  "Culture" => d.user.culture,
+	  "TwoFactorEnabled" => d.user.totp_secret,
+	  "Key" => d.user.key,
+	  "PrivateKey" => d.user.private_key,
+	  "SecurityStamp" => d.user.security_stamp,
+	  "Organizations" => "[]",
+	  "Object" => "profile",
+	}.to_json
+	
+
+  end
+  
+  # Used by the web vaul to connect and load the user profile/datas
+  get "/accounts/profile" do
+    content_type :json
+	response['access-control-allow-origin'] = '*'
+    d = device_from_bearer
+    if !d
+      return validation_error("invalid bearer")
+    end
+
+    {
+	  "Id" => d.user_uuid,
+	  "Name" => d.user.name,
+	  "Email" => d.user.email,
+	  "EmailVerified" => d.user.email_verified,
+	  "Premium" => d.user.premium,
+	  "MasterPasswordHint" => d.user.password_hint,
+	  "Culture" => d.user.culture,
+	  "TwoFactorEnabled" => d.user.totp_secret,
+	  "Key" => d.user.key,
+	  "PrivateKey" => d.user.private_key,
+	  "SecurityStamp" => d.user.security_stamp,
+	  "Organizations" => "[]",
+	  "Object" => "profile",
+	}.to_json
+  end
+
+  # Used to update masterpassword
+  post "/accounts/password" do
+	content_type :json
+	response['access-control-allow-origin'] = '*'
+    d = device_from_bearer
+    if !d
+      return validation_error("invalid bearer")
+    end
+	
+	need_params(:key, :masterpasswordhash, :newmasterpasswordhash) do |p|
+	  return validation_error("#{p} cannot be blank")
+	end
+
+	if !params[:key].to_s.match(/^0\..+\|.+/)
+	  return validation_error("Invalid key")
+	end
+	
+	begin
+	  Bitwarden::CipherString.parse(params[:key])
+	rescue Bitwarden::InvalidCipherString
+	  return validation_error("Invalid key")
+	end
+	
+	if d.user.password_hash == params[:masterpasswordhash]
+      d.user.key=params[:key]
+	  d.user.password_hash=params[:newmasterpasswordhash]
+	else
+	  return validation_error("Wrong current password")
+	end
+	  
+    User.transaction do
+	  if !d.user.save
+	    return validation_error("Unknown error")
+	  end
+	end
+	""
+  end
+
+  # Used to update email 
+  post "/accounts/email-token" do
+	content_type :json
+	response['access-control-allow-origin'] = '*'
+	validation_error("Not implemented yet")
+  end
+
   #
   # ciphers
   #
 
+  # Import from keepass or others via web vault
+  post "/ciphers/import" do
+    content_type :json
+	response['access-control-allow-origin'] = '*'
+
+	d = device_from_bearer
+    if !d
+      return validation_error("invalid bearer")
+    end
+
+	return validation_error("import tool not implemented yet")
+
+  end
```

Comment:
What does this function do?

Reply:
Currently, nothing. But it needs to be implemented to do the "mass import". Like importing from 1password, lastpass, or Keepass.
The web vault manage to parse all the formats, but it send the parameters via POST command. I couldn't determine the link between items and folders. So i didn't implemented it.